_posts/2024-04-20-coroutines.md

```diff
@@ -0,0 +1,348 @@
+
+From this perspective, coroutines are a third way of approaching these processes, which is not better than 1) or 2), but also not worse.
+Coroutines bring in a new (and not yet finetuned) mechanism that makes it possible to write an exchange of data over UART in a way that does not require as complex code as 1), does not suffer so many concurrency issues as 2) (and does not need its own stack, just frame).
+
+However, coroutines still suffer from the issue that one step of computation might take longer than expected, and we can assume that any errors in coroutines might be harder to inspect.
+
+What is interesting, is that coroutines have good potential to have more efficient context switches than threads, which was shown in this article: https://www.researchgate.net/publication/339221041_C20_Coroutines_on_Microcontrollers_-What_We_Learned
+
+Note that the article does not compare coroutines in a variety of situations, but it still shows something.
+
+### Problems
+
+There are multiple potential issues with coroutines and using them correctly, but I think there are a few that are noteworthy for the embedded community.
+
+#### Dynamic memory
+
+The first thing that I assume most of you noticed at the beginning is the fact the coroutine requires dynamic memory.
+This is not favourable in embedded in many cases, but there are ways to avoid that.
+
+Coroutines have `allocator` support, we can provide the coroutine with an allocator that can be used to get memory for the coroutine and hence avoid the dynamic allocation. (Approach that I can suggest)
+This is done by implementing custom `operator new` and `operator delete` on the `promise_type` which allocates the `entire frame`.
+
+An alternative is to rely on `halo` optimization (Heap Allocation Elision Optimization) if the coroutine is implemented correctly and the parent function executes the entire coroutine in its context.
+The compiler can optimize away the dynamic allocation and just store the frame on the stack of the coroutine's parent.
+This can be enforced by deleting appropriate `operator new` and `operator delete` overloads of the `promise_type`, but it seems clumsy to me.
+
+And to kinda ruin the pretty thing here, there is one catch. As of now, the compiler can't tell you how much memory the coroutine needs, as it is known only during the link time - the size of the coroutine frame can't be compiler constant. (TODO: link to the source for this)
+This means that you effectively can't prepare static buffers for the coroutines.
+
+What I would suggest (that is what I do), is to use coroutines for the long-term process and just build them during the initialization of the device.
+Or just live with the dynamic memory.
+
+(Note: if I could dream a bit here, I would want an explicit way of forcing the coroutine to live on the stack, which would have clearly defined behaviour of when and how it should happen, and with proper compiler errors if one fails to do so)
```

Comment:
Yeah, this would be the preferred way :(. It might help to flesh out the `halo` optimization more and why it isn't as ideal, since by your description it looks to accomplish this.

Reply:
I will ok this more to make sure it is explained properly, `halo` is optimization... you can't force compiler to do it, comiler usually won't tell you why it did it or did not do it, or it won't even tell you how exactly it decides